sparse/coo.py

```diff
@@ -2149,101 +2132,153 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
+    data = [arg.data[idx] for arg, idx in zip(args, sorted_indices)]
 
-def _reverse_self_other(func):
-    def wrapper(*args, **kwargs):
-        return func(args[1], args[0], *args[2:], **kwargs)
+    # Find matches between self.coords and other.coords
+    matched, matched_indices = _nary_match(*reduced_linear)
 
-    return wrapper
+    matched_coords = [c[:, idx] for c, idx in zip(coords, matched_indices)]
 
+    # Add the matched part.
+    matched_coords = _get_matching_coords(matched_coords, broadcast_params, matched_shape)
+    matched_data = [d[idx] for d, idx in zip(data, matched_indices)]
+
+    matched_indices = [sidx[midx] for sidx, midx in zip(sorted_indices, matched_indices)]
+
+    matched_arrays = [COO(matched_coords, d, shape=matched_shape) for d in matched_data]
 
-def _get_unmatched_coords_data(coords, data, shape, result_shape, matched_idx,
-                               matched_coords):
+    return matched_indices, matched_arrays
+
+
+def _unmatch_coo(func, args, mask, **kwargs):
     """
-    Get the unmatched coordinates and data - both those that are unmatched with
-    any point of the other data as well as those which are added because of
-    broadcasting.
+    Matches the coordinates for any number of input :obj:`COO` arrays.
+
+    First computes the matches, then filters out the non-matches.
 
     Parameters
     ----------
-    coords : np.ndarray
-        The coordinates to get the unmatched coordinates from.
-    data : np.ndarray
-        The data corresponding to these coordinates.
-    shape : tuple[int]
-        The shape corresponding to these coordinates.
-    result_shape : tuple[int]
-        The result broadcasting shape.
-    matched_idx : np.ndarray
-        The indices into the coords array where it matches with the other array.
-    matched_coords : np.ndarray
-        The overall coordinates that match from both arrays.
+    args : tuple[COO]
+        The input :obj:`COO` arrays.
 
     Returns
     -------
-    coords_list : list[np.ndarray]
-        The list of unmatched/broadcasting coordinates.
-    data_list : list[np.ndarray]
-        The data corresponding to the coordinates.
+    matched_idx : list[ndarray]
+        The indices of matched elements.
+    matched_coords : list[ndarray]
+        The matched coordinates.
+    matched_data : list[ndarray]
+        The matched data.
     """
-    params = _get_broadcast_parameters(shape, result_shape)
-    matched = np.zeros(len(data), dtype=np.bool)
-    matched[matched_idx] = True
-    unmatched = ~matched
-    data_zero = _zero_of_dtype(data.dtype)
-    nonzero = data != data_zero
+    matched_args = [a for a, m in zip(args, mask) if m]
+    unmatched_args = [a for a, m in zip(args, mask) if not m]
 
-    unmatched &= nonzero
-    matched &= nonzero
+    matched_arrays = _match_coo(*matched_args)[1]
 
-    coords_list = []
-    data_list = []
+    pos, = np.where([not m for m in mask])
```

Comment:
maybe use `np.flatnonzero()`?

Reply:
This isn't really a numerical operation. I've converted it to a `tuple(generator comprehension)` form and avoided `np.where` altogether. The exact code is
```
pos = tuple(i for i, m in enumerate(mask) if not m)
```